File: wisefy/src/androidTest/java/com/isupatches/wisefy/test/EnableWifiTests.java
package com.isupatches.wisefy.test;


import com.isupatches.wisefy.callbacks.EnableWifiCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import com.isupatches.wisefy.test.base.BaseTestClass;
import com.isupatches.wisefy.test.base.TestActivity;
import org.junit.Test;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class EnableWifiTests extends BaseTestClass<TestActivity> {

    public EnableWifiTests() {
        super(TestActivity.class);
    }

    @Test
    public void noCallbacks_failure() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        boolean result = mWiseFy.enableWifi();
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        assertEquals(false, result);
    }

    @Test
    public void noCallbacks_failure_nullWifiManager() {
        setManagersToNull();
        boolean result = mWiseFy.enableWifi();
        assertEquals(false, result);
    }

    @Test
    public void noCallbacks_success() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        boolean result = mWiseFy.enableWifi();
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        assertEquals(true, result);
    }

    @Test
    public void callbacks_failure() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(enableWifiCallbacks);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        verify(enableWifiCallbacks, timeout(2000)).failureEnablingWifi();
    }

    @Test
    public void callbacks_failure_nullCallback() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(false);
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(null);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        verify(enableWifiCallbacks, never()).failureEnablingWifi();
    }

    @Test
    public void callbacks_failure_nullWifiManager() {
        setManagersToNull();
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(enableWifiCallbacks);
        verify(mMockWiFiManager, never()).setWifiEnabled(true);
        verify(enableWifiCallbacks, timeout(2000)).enableWifiWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
    }

    @Test
    public void callbacks_failure_nullWifiManager_nullCallback() {
        setManagersToNull();
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(null);
        verify(mMockWiFiManager, never()).setWifiEnabled(true);
        verify(enableWifiCallbacks, never()).enableWifiWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
    }

    @Test
    public void callbacks_success() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(enableWifiCallbacks);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        verify(enableWifiCallbacks, timeout(2000)).wifiEnabled();
    }

    @Test
    public void callbacks_success_nullCallback() {
        setManagers();
        when(mMockWiFiManager.setWifiEnabled(anyBoolean())).thenReturn(true);
        EnableWifiCallbacks enableWifiCallbacks = mock(EnableWifiCallbacks.class);
        mWiseFy.enableWifi(null);
        verify(mMockWiFiManager, timeout(2000)).setWifiEnabled(true);
        verify(enableWifiCallbacks, never()).wifiEnabled();
    }
}

File: wisefy/src/test/java/com/isupatches/wisefy/EnableWifiTests.kt
package com.isupatches.wisefy

import android.os.Build

import com.isupatches.wisefy.callbacks.EnableWifiCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import com.isupatches.wisefy.internal.base.BaseUnitTest

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Tests the ability to enable a device's wifi.
 *
 * @author Patches
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.M])
internal class EnableWifiTests : BaseUnitTest() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.enableWifi_failure()
        assertEquals(false, wisefy.enableWifi())
        verificationUtil.didNotTryToEnableWifi()
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_failure()
        assertEquals(false, wisefy.enableWifi())
        verificationUtil.triedToEnableWifi()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_success()
        assertEquals(true, wisefy.enableWifi())
        verificationUtil.triedToEnableWifi()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.enableWifi_failure()
        val mockCallbacks = mock(EnableWifiCallbacks::class.java)
        wisefy.enableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToEnableWifi()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.enableWifi_failure()
        nullCallbackUtil.callEnableWifi()
        verificationUtil.didNotTryToEnableWifi()
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_failure()
        val mockCallbacks = mock(EnableWifiCallbacks::class.java)
        wisefy.enableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureEnablingWifi()
        verificationUtil.triedToEnableWifi()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_failure()
        nullCallbackUtil.callEnableWifi()
        verificationUtil.triedToEnableWifi()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_success()
        val mockCallbacks = mock(EnableWifiCallbacks::class.java)
        wisefy.enableWifi(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wifiEnabled()
        verificationUtil.triedToEnableWifi()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.enableWifi_success()
        mockNetworkUtil.enableWifi_success()
        nullCallbackUtil.callEnableWifi()
        verificationUtil.triedToEnableWifi()
    }
}
